docs(adr): Subspace fee grant (#1042)
desmos-labs/desmos: docs/architecture/adr-017-subspace-fee-grant.md

# ADR 017: Subspace fee grant

## Changelog

- Nov 14th, 2022: First draft

## Status

PROPOSED

## Abstract

This ADR introduces a new subspace-based fee grant method, which allows subspace service providers to pay the fees of subspace related transactions for the users.

## Context

Currently, one of the major problems of current Web3 services is complicated to use, it requires users to have tokens paying transaction fees so that they can use the service. For instance, in order to create a post on Desmos-based social network, users have to understand:
- what `transaction` is and why they need some DSM to broadcast it;
- get some DSM either via an on-ramp or by swapping existing funds.

> The `x/feegrant` gives the possibility to pay fees for the users, meaning that users can use the service without understanding how Web3 works behind. However, the `x/feegrant` allowance is not subspace-specified so subspace fees providers might unexpected pay the fees of the transactions inside another subspace.

Reviewer: I think another thing we should allow is granting allowances to **groups** of users. This way a subspace owner would not need to: 
1. issue an amount of grants equals to the number of active users 
2. remove feegrant allowances as soon as a user becomes inactive/is banned/etc.

Author: The idea is so nice, it is so useful for the subspace owner to manage the allowance of each several groups.

Author: Using group allowance will have a problem that the fees provider can't set spend and periodic limit for each users since the recorded target is group id. It means that all the users in a group shares the same granted amount if the limit is set.

To solve this problem, Desmos needs to create spend records for all users when a group is granted. Besides, after executing `AddUserToGroup`, `RemoveUserFromGroup`, `DeleteGroup` and `GrantGroupAllowance` actions, desmos are required to update the spend records for all users.

Reviewer: > Using group allowance will have a problem that the fees provider can't set spend and periodic limit for each users since the recorded target is group id. It means that all the users in a group shares the same granted amount if the limit is set.

I don't think this is necessarily a problem. It's just how it's supposed to work: if you set a group limit, the limit is shared among the overall group. What owners can do is setting a time limit, but not an amount limit. Or, simply, they can set a very high amount limit instead. That being said, I think it's still something we should probide since it would simplify the grant management a lot. 

> In addition, a user might have several allowances by joining different groups, it also increases the complexity.

I don't think it's extremely complicated for the following reasons: 
1. user groups will not be that many inside a subspace.  
   I don't see a reason why a subspace would have hundreds or thousands of user groups: it just makes it really hard for the owner to manage them properly. 
2. even if the subspace has millions of groups, it doesn't matter. We will iterate *first* on the grants given to various groups, and then check if either one of those groups contains the user. So we scale based on the number of grants, instead of on the number of groups. And, since groups are hierarchical (like permissions), we can stop as soon as we find the first group

> Any thoughts?

Author: Got you! The complicated thingy only happens if desmos records all users in the granted group but it is not necessarily. The ADR group allowance will be all the users share the group limit.

1. the user asks fees provider a fee grant inside the specified subspace;
2. the fees provider send a subspace fee grant transaction to Desmos;
3. Desmos executes the subspace fee grant transaction successfully;
4. the user can use the service without any tokens inside the specified subspace, the fees will be paid by the fees provider.

```mermaid
sequenceDiagram
    participant User
    participant Fees Provider
    participant Desmos
    
    User->>Fees Provider: 1. ask for a subspace fee grant
    Fees Provider->>Desmos: 2. send subspace fee grant tx
    Desmos->>Desmos: 3. execute subspace fee grant tx
    Desmos-->>Fees Provider:  return ok
    Fees Provider-->>User: return ok

    User->>Desmos: 4. use app with subspace tx
    Desmos-->>User: return ok
```

### DeductFeeDecorator

Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting fees from the signer/feepayer of a transaction. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.

The new subspace-specified `DeductFeeDecorator` will operate the fees with the process as follows:

> 1. check all the messages in the transaction are the subspace messages and all of them are to the same subspace;

Reviewer: With this method, could a subspace owner issue a grant for both `MsgSaveProfile` and `MsgCreatePost` messages? If so, could then a user perform a transaction containing one instance of both, without any issue? 

From what I read it seems that these kind of grants are related only to subspace messages, but I think it would be better to allow subspace owners to also issue grants related to `MsgSaveProfile` or other kind of messages since those might be useful when creating a Desmos Profile for the first time within an application.

> 1. check all the messages in the transaction are the subspace messages and all of them are to the same subspace;

Reviewer: With this method, could a subspace owner issue a grant for both `MsgSaveProfile` and `MsgCreatePost` messages? If so, could then a user perform a transaction containing one instance of both, without any issue? 

From what I read it seems that these kind of grants are related only to subspace messages, but I think it would be better to allow subspace owners to also issue grants related to `MsgSaveProfile` or other kind of messages since those might be useful when creating a Desmos Profile for the first time within an application.

Author: `MsgCreatePost` is allowed, but `MsgSaveProfile` is not since `x/profiles` should be the root module of all desmos modules, additionally, `x/profile` messages are not subspace-specified. We should exclude the `x/profiles` messages from this ADR.

In order to pay the fees of other non-subspace messages like `MsgSaveProfile` for the user in this scenario, the subspace owner can use `x/feegrant` to grant a filtered allowance including `MsgSaveProfile` with spend and periodic limit.

Author: I realize that instead of checking if all the messages are the same subspace messages or not, we can specified the subspace id and any needed info in the `memo` field. A format like `subspace-1-0` to use the subspace 1 group 0(default) to check the fee payer.

Reviewer: > I realize that instead of checking if all the messages are the same subspace messages or not, we can specified the subspace id and any needed info inside the `memo` field. It would be like a format `subspace-1` to check the fee payer and grant on subspace 1. Then, we can grant any other non-subspace messages. What do you think?

I don't think that's a viable and scalable solution. It would cause the memo field to become unusable for other purposes, limiting the DApps and requiring them to know more things before broadcasting the messages. I think it's better to just check the transaction messages instead

Author: Another idea is to wrap the sdk.Tx to desmos.Tx containing the `subspace_id` and `group_id` soo that we can use these field to get the subspace grant easily in the deducting transaction phase. But it will require a new transaction type to increase the maintaining cost.

## Decision

We will implement a subspace-specified fee grant process based on `x/feegrant` that allows subspace fees providers to pay fees for the users inside the specified subspace. This system will allow subspace owners/admins to issue a fee grant towards either a users group or a single user, so that they can later leverage this grant to execute subspace-related transactions without having to worry about paying fees.

### DeductFeeDecorator

Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting the fees of a transaction from the signer/feepayer. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.

The new subspace-specified `DeductFeeDecorator` will work as follows:
1. check if all the messages in the transaction are messages related to the same subspace;

> 2. if the transaction contains subspace messages from the same subspace and a fee grant exists, run the `x/subspaces` `DeductFeeDecorator`; otherwise run the `x/auth` `DeductFeeDecorator`,

Author: Yes, exactly.

> Currently, `x/auth` provides a `DeductFeeDecorator` based on `x/feegrant` to execute the action deducting the fees of a transaction from the signer/feepayer. We will build a new subspace-specified `DeductFeeDecorator` to replace the current one.
> 
> The new subspace-specified `DeductFeeDecorator` will work as follows:
> 1. check if all the messages in the transaction are messages related to the same subspace;
> 2. if the messages in the transaction are all realted to the same subspace and a fee grant exists for that subspace, run the `x/subspaces` `DeductFeeDecorator`; otherwise run the `x/auth` `DeductFeeDecorator`,

Reviewer: ```suggestion
Currently, the `x/auth` module provides a `DeductFeeDecorator` that deducts the fees of a transaction from the signer/feepayer. We will build a new subspace-specific `DeductFeeDecorator` based on the existing one that acts as follows:

1. check if all the messages in the transaction are messages related to the same subspace;
2. perform the following operation:
   1. if all the messages are all realted to the same subspace and a fee grant exists for the fee payer inside the subspace, run the `x/subspaces` `DeductFeeDecorator`
   2. otherwise, run the `x/auth` `DeductFeeDecorator`
```

### Types

#### Allowance

> The `x/feegrant` module already provides an interface (`FeeAllowanceI`) to represent a generic allowance, along with a set of useful implementations like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Since we are not going to introduce new kinds of allowances, nor we are going to edit how an allowance is represented, we will reuse these types.

Reviewer: I was thinking about one thing: should we creae a new implementation of `FeeAllowanceI` named `SubspaceFeeAllowance` that is related to a single subspace? This way, we could even leverage the default `x/feegrant` `MsgGrantFeeAllowance` message to create such allowances.

Author: Using the existing ones is enough from my side. 
These cases would be covered:
1. granter grants all the subspace messages with a spend limit by `BasicAllowance`
2. granter grants some of subspace messages with a spend limit by `AllowedMsgAllowance`
3. granter grant all the subspaces messages with periodic limit by `PeriodicAllowance` with `BasicAllowance`
4. granter grant all some of subspaces messages with periodic limit by `PeriodicAllowance` with `AllowedMsgAllowance`

Reviewer: @dadamu I see. my idea was to remove the need of handling user-specific grants inside the `x/subspace` module, and instead implement a new `FeeAllowanceI` that can be properly handled by the `x/feegrant` module. Maybe the same could be done for the group as well? This was we don't have to store anything using custom keys but we can simply reply on the `x/feegant` module to store everything. But I'm not sure whether that's possible

Author: @RiccardoM 
The subspace group grant is impossible since `feegrant` only gives the possibility to user grant like [this](https://github.com/cosmos/cosmos-sdk/blob/main/x/feegrant/keeper/keeper.go#L43), grantee must be a `sdk.AccAddress`.
On the other hand, if we move subspace user grant into `x/feegrant`, the check would be no longer subspaces-specific, in addition, it would conflict with the non-subspace feegrant.

Reviewer: @dadamu Ok, perfect. We can leave it like you designed it then

#### User Grant

The `x/feegrant` module provides a `Grant` object used to store the `granter`, `grantee` and what kind of `allowance` is granted to a specific user. Since this is enough information for us as well, we will use the same object to store user-related fee grants.

#### Group Grant

In order to represent a fee grant granted to a user group, we will implement a `GroupGrant` type. This will contain the `granter`, `group_id` and what kind of `allowance` is granted to the group:

```protobuf
message GroupGrant {
    // granter is the address of the user granting an allowance of their funds.
  string granter = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];

  // the id of the group being granted an allowance of another user's funds.
  uint32 group_id = 2;

  // allowance can be any of basic, periodic, allowed fee allowance.
  google.protobuf.Any allowance = 3 [(cosmos_proto.accepts_interface) = "FeeAllowanceI"];
}
```

### Store

#### Group grant

> In order to simplify granters to manage the group allowance, the granted group allowance will be stored in the keys having the structure like:

Reviewer: ```suggestion
In order to simplify the management of group allowances, each granted group allowance will be stored usign the following key:
```

```
SubspaceGroupAllowancePrefix | SubspaceID | GranterAddress | GroupID | -> Protobuf(GroupGrant)
```

> This structure allows granters to easily manage the group allowance inside a subspace by iterating over all allowances for the granters, which will be the most used query. In the other hand, grantees must know who the granter is when using the application, they can find their grant with O(N) time complexity, N is the number of granted groups by the granter.

Reviewer: ```suggestion
This structure allows granters to easily manage the group allowance inside a subspace by iterating over all allowances for the granters, which will be the most used query. On the other hand, it allows to perform the following queries: 
1. finding a specific grant given the granter address.   
   This can be done in O(N), where N is the number of granted groups by the granter.
2. finding all the grans given to a user based on the groups they are part of.
   This can be done in O(N * M), where N is the number of granters, and M is the number of grants each granter has given.

Considering that: 
- only users with permissions will be able to create grants in each subspace
- there will be a small number of grants each granter will create

then the time complexity of searching if a user has been granted a fee grant should not become too large.
```

#### User grant

> Each subspace fee granted allowance will be stored in the keys having the structure as follows:

Reviewer: ```suggestion
Each subspace fee granted allowance will be stored using keys with the following structure:
```

```
SubspaceUserAllowancePrefix | SubspaceID | GranterAddress | GranteeAddress |-> Protobuf(Grant)
```

> This structure allows granters to easily manage their grants inside a subspace by iterating over all grants for the granters, which will be the most used query. In the other hand, grantees must know who the granter is when using the application, they can directly find their grant with O(1) time complexity.

Reviewer: ```suggestion
This structure allows granters to easily manage their grants inside a subspace by iterating over all grants for the granters, which will be the most used query. On the other hand, it allows the following queries:

1. retrieving a specific grant.  
   This will require grantees to know who the granter address, and can be done in O(1).
2. retrieving all the grants that a user has been granted. 
   This can be done by iterating over all the granters' grants in O(N * M), where N is the number of granters and M is the number of grants each granter has granted.
```

Reviewer: @dadamu But what if the user does not know the granter address, or wants the overall grants that have been granted? In that case we need to iterate over the prefix `SubspaceUserAllowancePrefix | SubspaceID` and check if each grant is referring to the user or not

Author: If the user does not know the granter, it would be O(N), where N is the number of all the grants.
However, from my side, the grantee must know who is the granter. Assume that there is an application provided by a granter, the granter would always grant new users subspace allowance, then users use that application to operate inside the subspace must know who the application granter.

Author: @RiccardoM If the user does not know the granter, it would be O(N), where N is the number of all the grants.
However, from my side, the grantee must know who is the granter. Assume that there is an application provided by a granter, the granter would always grant new users subspace allowance, then users use that application to operate inside the subspace must know who the application granter.

Reviewer: @dadamu Ok, yeah O(N) with N number of grants looks decent anyway. 

> then users use that application to operate inside the subspace must know who the application granter

Not necessarily, there might be multiple admins that are allowed to create a user-group fee grant. In that case, the user would not not the granter. Or they might have multiple grants because they are inside multiple groups. I think we should not make it necessary for users to know their granter. Also becuase this is a query that does not consume gas and can be implemented quite easily. I don't see any harm in allowing it

    uint64 subspace_id = 1;

    // the id of the group being granted an allowance of another user's funds.  
    uint32 group_id = 2;

    // the address of the user granting an allowance of their funds.
    string granter = 3;

    // allowance can be any of fee allowances which implements FeeAllowanceI.
    google.protobuf.Any allowance = 4 [(cosmos_proto.accepts_interface) = "FeeAllowanceI"];
}

// MsgGrantGroupAllowanceResponse defines the Msg/GrantAllowanceResponse response type.
message MsgGrantGroupAllowanceResponse {}

// MsgRevokeGroupAllowance removes any existing allowance from granter to the group inside the subspace.
message MsgRevokeGroupAllowance {
    // the id of the subspace where the granter grants the allowance to the group.
    uint64 subspace_id = 1;

    // the id of the group being granted an allowance of another user's funds.
    uint32 group_id = 2;

    // the address of the user granting an allowance of their funds.
    string granter = 3;
}

// MsgRevokeGroupAllowanceResponse defines the Msg/RevokeAllowanceResponse response type.
message MsgRevokeUserAllowanceResponse {}
```

### `Query` Service

> In order to allow clients to easily query for allowances we will implement a new queries:

Reviewer: ```suggestion
In order to allow clients to easily query for allowances we will implement the following new queries:
```

```protobuf
service Query {
    // UserAllowances returns all the grants for users.
    rpc UserAllowances(QueryUserAllowancesRequest) returns (QueryAllowancesResponse) {

>         option (google.api.http).get = "/desmos/subspaces/v3/subspaces/{subspace_id}/granter/{granter}/users/allowances";

Reviewer: I think the `granter` can be optional, as we should also allow queries: 
- all the grants in a subspace 
- all the grants given a granter address (but not grantee)

```suggestion
        option (google.api.http).get = "/desmos/subspaces/v3/subspaces/{subspace_id}/allowances/users";
```

    }
    // GroupAllowances returns all the grants for groups.
    rpc GroupAllowances(QueryGroupAllowancesRequest) returns(QueryGroupAllowancesResponse) {

>         option (google.api.http).get = "/desmos/subspaces/v3/subspaces/{subspace_id}/granter/{granter}/groups/allowances";

Reviewer: I think the `granter` can be optional, as we should also allow queries:

- all the grants in a subspace
- all the grants given a group id (but not grantee)

```suggestion
        option (google.api.http).get = "/desmos/subspaces/v3/subspaces/{subspace_id}/allowances/groups";
```

Author: > all the grants given a group id (but not grantee)
I guess you make typos, it should be "all the grants given a granter (but not group id)"

    }
}

// QueryUserAllowancesRequest is the request type for the Query/UserAllowances RPC method.
message QueryUserAllowancesRequest {
    // the id of the subspace where the granter grants the allowance to the grantee.
    uint64 subspace_id = 1;

>     // the address of the user granting an allowance of their funds.

Reviewer: We should make it optional to allow queries all the grants present, and all the grants given a grantee address
```suggestion
    // (optional) the address of the user granting an allowance of their funds.
```

    string granter = 2;

    // (Optional) the address of the user being granted an allowance of another user's funds.
    string grantee = 3;

    // pagination defines an pagination for the request.
    cosmos.base.query.v1beta1.PageRequest pagination = 4;
}

// QueryUserAllowancesResponse is the response type for the Query/UserAllowances RPC method.
message QueryUserAllowancesResponse {
    // allowances are allowance's granted for grantee by granter.
    repeated cosmos.feegrant.v1beta1.Grant allowances = 1;

    // pagination defines an pagination for the response.
    cosmos.base.query.v1beta1.PageResponse pagination = 2;
}

// QueryGroupAllowancesRequest is the request type for the Query/GroupAllowances RPC method.
message QueryGroupAllowancesRequest {
    // the id of the subspace where the granter grants the allowance to the grantee.
    uint64 subspace_id = 1;

>     // the address of the user granting an allowance of their funds.

Reviewer: We should make this optional to allow quering all the grants, and all the grants given a group id 
```suggestion
    // (optional) the address of the user granting an allowance of their funds.
```

    string granter = 2;

    // (Optional) the id of the group being granted an allowance of another user's funds.
    uint32 group_id = 3;

    // pagination defines an pagination for the request.
    cosmos.base.query.v1beta1.PageRequest pagination = 4;
}

// QueryGroupAllowancesResponse is the response type for the Query/GroupAllowances RPC method.
message QueryGroupAllowancesResponse {
    // allowances are allowance's granted for grantee by granter.
    repeated cosmos.subspace.v3.GroupGrant allowances = 1;

    // pagination defines an pagination for the response.
    cosmos.base.query.v1beta1.PageResponse pagination = 2;
}
```

## Consequences

### Backwards Compatibility

> The proposed solution introduces a new set of store keys, it is completely backward compatible.

Reviewer: ```suggestion
The proposed solution introduces a new set of store keys, thus it is completely backward compatible.
```

### Positive

- Allow subspace fee providers to grant fees allowances to their users.

### Negative

- Storing extra subspace grant info takes up more storage space;

> - Having more check in the deducting fees phase slow down the transaction operation;

Reviewer: ```suggestion
- Having more checks during the fee deducton phase slows down the transaction processing operation;
```

> - Removing expired allowances at the block begin phase increases the block production time.  

Reviewer: Is this necessary? Does the Cosmos SDK remove the fee grants at each block, or do they remove them when they are tried to be executed, if no longer valid?

Author: Cosmos SDK remove all the expired fee grants every block. We can make it removed when the user try to execute with feegrant, but it will leave invalid allowances in the database. I prefer to remove expired allowances every blocks.

Reviewer: Ok, then I think we should add this inside the ADR since there's not reference to a `BeginBlock` implementation

### Neutral

- Not known

## References

> https://docs.cosmos.network/v0.46/modules/feegrant/01_concepts.html

Reviewer: ```suggestion
- [Cosmos SDK fee grant concepts](https://docs.cosmos.network/v0.46/modules/feegrant/01_concepts.html)
```

- there will be a small number of grants each granter will create

then the time complexity of searching if a user has been granted a fee grant should not become too large.

#### User grant

Each subspace fee granted allowance will be stored using keys with the following structure:
```
SubspaceUserAllowancePrefix | SubspaceID | GranterAddress | GranteeAddress |-> Protobuf(Grant)
```

This structure allows granters to easily manage their grants inside a subspace by iterating over all grants for the granters, which will be the most used query. On the other hand, it allows the following queries:
1. retrieving a specific grant.  
   This will require grantees to know who the granter address, and can be done in O(1).
2. retrieving all the grants that a user has been granted. 
   This can be done by iterating over all the granters' grants in O(N), where N is the number of all the grants inside the given subspace.

### Expired Allowance

The subspace owners may want to set the allowance to expire after a certain amount of time. In order to reduce the database size, we will use a custom `BeginBlocker` to remove all expired allowances.

#### Store

> In order to make it easier to remove expired allowances, we will implement queues that are sorted by expiration time. Then, during the BeginBlock phase, Desmos can iterate over all expired keys before the block time and remove them. The structure will be like:

Reviewer: ```suggestion
In order to make it easier to remove expired allowances, we will implement queues sorted by expiration time. Then, during the `BeginBlock` phase, we iterate over all expired keys before the block time and remove them. The structure of the queues keys will be the following:
```

The new subspace-specified `DeductFeeDecorator` will deal with the fees with the process as follows:
1. check all the messages in the transaction are the subspace messages and all of them are to the same subspace;
2. apply `x/subspaces` `DeductFeeDecorator` if the transaction contains subspace messages from the same subspace and the user/group fee grant exists, or apply `x/auth` `DeductFeeDecorator`;
3. deduct the fees from the fee payer.

```mermaid
graph TD
  id1([Start]) --> id2[Check transaction]
  id2 --> id3{Are all subspace msgs<br/ >to the same subspace?}
  id3 -- YES --> id4{Is the user/group grant exist?}
  id3 -- NO --> id5[Apply `x/auth`<br />DeductFeeDecorator]
  id4 -- YES --> id6[Apply `x/subspaces`<br />DeductFeeDecorator]
  id4 -- NO --> id5
  id5 --> id7[Deduct fees]
  id6 --> id7
  id7 --> id8[End]
```

### Types

#### Allowance

Currently, `x/feegrant` provides an interface `FeeAllowanceI` and the implementations for the interface like `BasicAllowance`, `AllowedMsgAllowance` and `PeriodicAllowance`. Each of them already have the operation process to deal with the allowance. We will reuse them to be the allowance object inside the `x/subspace`.

> #### Grant
> 
> As `Allowance`, `x/feegrant` has `Grant` object to record all full context, which contains `granter`, `grantee` and what kind of `allowance` is granted. We will reuse it to store in the subspace namespace KVStore.

Reviewer: ```suggestion
#### User Grant

The `x/feegrant` module provides a `Grant` object used to store the `granter`, `grantee` and what kind of `allowance` is granted to a specific user. Since this is enough information for us as well, we will use the same object to store user-related fee grants.
```

#### GroupGrant

> To show the group grant, we will implement the `GroupGrant` object to record all full context containing `granter`, `group_id` and what kind of `allowance` is granted, the structure will be as follows:

Reviewer: ```suggestion
#### Group Grant

In order to represent a fee grant granted to a user group, we will implement a `GroupGrant` type. This will contain the `granter`, `group_id` and what kind of `allowance` is granted to the group:

```
